arithmetization/src/main/java/net/consensys/linea/blockcapture/reapers/Reaper.java

```diff
@@ -39,34 +43,79 @@
  */
 public class Reaper {
-  /** Collect the reads from the state */
-  private final StorageReaper storage = new StorageReaper();
+  /** Collect storage locations read / written by the entire conflation */
+  private final StorageReaper conflationStorage = new StorageReaper();
 
-  /** Collect the addresses read from the state */
-  private final AddressReaper addresses = new AddressReaper();
+  /** Collect the account address read / written by the entire conflation */
+  private final AddressReaper conflationAddresses = new AddressReaper();
 
   /** Collect the blocks within a conflation */
   private final List<BlockSnapshot> blocks = new ArrayList<>();
 
+  /**
+   * Transaction index is used to do determine the current transaction being processed within the
+   * current block.
+   */
+  private int txIndex;
+
+  /** Collect storage locations read / written by the current transaction */
+  private StorageReaper txStorage = new StorageReaper();
+
+  /** Collect the account address read / written by the current transaction */
+  private AddressReaper txAddresses = new AddressReaper();
+
   public void enterBlock(final BlockHeader header, final BlockBody body) {
     this.blocks.add(
         BlockSnapshot.of((org.hyperledger.besu.ethereum.core.BlockHeader) header, body));
-    this.addresses.touch(header.getCoinbase());
+    this.conflationAddresses.touch(header.getCoinbase());
+    txIndex = 0; // reset
   }
 
   public void enterTransaction(Transaction tx) {
     this.touchAddress(tx.getSender());
     tx.getTo().ifPresent(this::touchAddress);
+    // Configure tx-local reapers
+    this.txStorage = new StorageReaper();
+    this.txAddresses = new AddressReaper();
   }
 
-  public void exitTransaction(boolean success) {}
+  public void exitTransaction(
+      WorldView world,
+      boolean status,
+      Bytes output,
+      List<Log> logs,
+      long gasUsed,
+      Set<Address> selfDestructs) {
+    // Identify relevant transaction snapshot
+    TransactionSnapshot txSnapshot = blocks.getLast().txs().get(txIndex);
+    // Convert logs into hex strings
+    List<String> logStrings = logs.stream().map(l -> l.getData().toHexString()).toList();
+    // Convert destructed account addresses into into hex strings
+    List<String> destructStrings = selfDestructs.stream().map(Address::toHexString).toList();
+    // Collapse accounts
+    final List<AccountSnapshot> accounts = this.txAddresses.collapse(world);
+    // Collapse storage
+    final List<StorageSnapshot> storage = txStorage.collapse(world);
+    // Construct the result snapshot
+    TransactionResultSnapshot resultSnapshot =
+        new TransactionResultSnapshot(
+            status, output.toHexString(), logStrings, gasUsed, accounts, storage, destructStrings);
+    // Record the outcome
+    txSnapshot.setTransactionResult(resultSnapshot);
+    // Reset for next transaction
+    txIndex++;
+    this.txStorage = new StorageReaper();
```

Comment:
Given we initialize them in the enterTransaction, I we shouldn't need to initialize them here. Should we instead add a close method here, and in the enter check that the reapers are either null or close? This way we would be sure that tx and entered then exited one after the other?

Reply:
Originally I did something like that, but I had a problem.  Let me revisit that.

Reply:
Ok, have resolved the problem here.  It was that, for some transactions, `traceStartTransaction()` is never called!  Instead, `tracePrepareTransaction()` is always called but we can go straight to `traceEndTransaction()` from there without calling `traceStartTransaction()`.

I've also refactored the way in which the `TransactionChecker` was integrated.  Previously, it was a wrapper around an existing `OperationTracer` (which was `ZkTracer`).  However, now its a standalone implementation and I use a separate mechanism for multiplexing tracers together (defined in `ConflationAwareOperationTracer`).  This has an additional benefit that I can relatively easily include the `BlockCapturer` for testing purposes.